### PR TITLE
libbpf-tools: add memleak

### PR DESCRIPTION
**Pull request by chenhengqi**
This is a work in progress.

### CODE REVIEW

**Comment by chenhengqi** [comment]
@anakryiko  I have problem building this tool, can you offer some advices ?

```
  BPF      memleak.bpf.o
  GEN-SKEL memleak.skel.h
libbpf: bad map relo against section 6
Error: failed to open BPF object file: Relocation failed
make: *** [Makefile:91: .output/memleak.skel.h] Error 255
make: *** Deleting file '.output/memleak.skel.h'
```

File: libbpf-tools/memleak.bpf.c [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+	if (addr) {
+		info.timestamp_ns = bpf_ktime_get_ns();
+		info.stack_id = bpf_get_stackid(ctx, &stack_traces, stack_flags);
+		bpf_map_update_elem(&allocs, &addr, &info, BPF_ANY);
+		incr_stat(info.stack_id, info.size);
+	}
+
+	if (debug_enabled) {
+		bpf_trace_printk("alloc exited, size = %lu, result = %lx\n", info.size, addr);
```
Comment (anakryiko) [inline libbpf-tools/memleak.bpf.c:159]: this is the problem, you can't just use strings directly with bpf_trace_printk(). Either use bpf_printk(), which is a macro that will put such string into on stack array, or you can make this string a static variable:

```
static const char[] fmt = "alloc exited, size = %lu, result = %lx\n";
```

**Review by anakryiko — CHANGES_REQUESTED** [review]

File: libbpf-tools/memleak.bpf.c (hunk 1 above)
Comment (chenhengqi) [reply libbpf-tools/memleak.bpf.c:159]: Ah, thanks.

**Comment by chenhengqi** [comment]
Closed in favor of #4502.

**Pull request closed without merging**